src/auth.rs

```diff
@@ -0,0 +1,180 @@
+use crate::config::Config;
+use crate::error::Error;
+
+use reqwest::header::AUTHORIZATION;
+use rocket::{
+    fairing::{AdHoc, Fairing},
+    http::{Cookie, CookieJar, SameSite},
+    outcome::IntoOutcome,
+    request::{self, FromRequest, Request},
+    response::Redirect,
+};
+use rocket_oauth2::{OAuth2, TokenResponse};
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct LoginUrl {
+    pub login_url: String,
+}
+
+#[derive(Debug, strum_macros::EnumString)]
+pub enum OauthProvider {
+    Google,
+    Microsoft,
+}
+
+pub struct TokenCookie(String);
+
+impl std::fmt::Display for TokenCookie {
```

Comment:
TokenCookie seems to contain a secret, should we really be implementing display (or debug for that matter) for it?